zebra-chain/src/diagnostic/task/thread.rs

```diff
@@ -0,0 +1,111 @@
+//! Diagnostic types and functions for Zebra OS thread tasks:
+//! - task handles
+//! - errors and panics
+
+use std::{
+    panic,
+    sync::Arc,
+    thread::{self, JoinHandle},
+};
+
+use super::{CheckForPanics, WaitForPanics};
+
+impl<T> CheckForPanics for thread::Result<T> {
+    type Output = T;
+
+    /// Waits for the thread to finish, then panics if the thread panicked.
+    ///
+    /// Threads can't be cancelled except by using a panic, so there are no thread errors here.
+    #[track_caller]
+    fn check_for_panics(self) -> Self::Output {
+        match self {
+            // The value returned by the thread when it finished.
+            Ok(thread_output) => thread_output,
+
+            // A thread error is always a panic.
+            Err(panic_payload) => panic::resume_unwind(panic_payload),
+        }
+    }
+}
+
+impl<T> WaitForPanics for JoinHandle<T> {
+    type Output = T;
+
+    /// Waits for the thread to finish, then panics if the thread panicked.
+    #[track_caller]
+    fn wait_for_panics(self) -> Self::Output {
+        self.join().check_for_panics()
+    }
+}
+
+impl<T> WaitForPanics for Arc<JoinHandle<T>> {
+    type Output = Option<T>;
+
+    /// If this is the final `Arc`, waits for the thread to finish, then panics if the thread
+    /// panicked. Otherwise, returns the thread's return value.
+    ///
+    /// If this is not the final `Arc`, drops the handle and immediately returns `None`.
+    fn wait_for_panics(self) -> Self::Output {
+        // If we are the last Arc with a reference to this handle,
+        // we can wait for it and propagate any panics.
+        //
+        // We use into_inner() because it guarantees that exactly one of the tasks gets the
+        // JoinHandle. try_unwrap() lets us keep the JoinHandle, but it can also miss panics.
+        //
+        // This is more readable as an expanded statement.
+        #[allow(clippy::manual_map)]
+        if let Some(handle) = Arc::into_inner(self) {
+            Some(handle.wait_for_panics())
+        } else {
+            None
+        }
+    }
+}
+
+impl<T> CheckForPanics for &mut Option<Arc<JoinHandle<T>>> {
+    type Output = Option<T>;
+
+    /// If this is the final `Arc`, checks if the thread has finished, then panics if the thread
+    /// panicked. Otherwise, returns the thread's return value.
+    ///
+    /// If the thread has not finished, or this is not the final `Arc`, returns `None`.
+    fn check_for_panics(self) -> Self::Output {
```

Comment:
Does this need `#[track_caller]` as well?

Reply:
They all do, thanks for catching that. Fixed in [b6a4f63](https://github.com/ZcashFoundation/zebra/pull/7219/commits/b6a4f631cd1daaf0fe8e2922a5753e36bbd60e81)